_posts/2018-12-18-debugging-bad-rows-on-gcp/index.md

```diff
@@ -0,0 +1,216 @@
+
+### 1. Create an external table and diagnose
+
+In the BigQuery UI, create a dataset to house your bad rows data - here I've named it `bad_rows`. Then create a table in that dataset, with the following options:
+
+**Create table from** Google Cloud Storage
+
+You'll want to point it to the root of your bad rows bucket with the `/*` operator, to encapsulate all bad rows. File format is JSON.
+
+**Table type** should be External table
+
+Name the table something sensible - I've gone with bad_rows_external.
+
+Finally, autodetect schema and input parameters.
+
+![Create external Table][create-external]
+
+Click create table and you're good to go. You can already start exploring your bad rows using SQL.
+
+If we take a look at the table schema, we'll see that there are three fields in the data - `failure_tstamp`, a nested errors object, containing `message` and `level`, and `line` - which is the base64 encoded payload containing the data.
+
+![Table Schema][table-schema]
+
+#### 1.1 Limiting Queries:
+
+Since our external table is built off the set of all of our bad data, we should take care to limit our queries appropriately when using it.
+
+**Note that with external tables, the BigQuery UI's validator isn't guaranteed to accurately reflect the amount of data your query scans - so vigilance is advised.**
+
+Bad rows are stored at time-partitioned filepaths, so we can trim the `_FILE_NAME` pseudo-column to a timestamp-format string and convert it to limit our queries. In standardSQL, this is done as follows:
+
+```SQL
+DATE(PARSE_TIMESTAMP('%Y-%m-%dT%H:%M:%S', LTRIM(REGEXP_EXTRACT(_FILE_NAME, 'output-[0-9]+-[0-9]+-[0-9]+T[0-9]+:[0-9]+:[0-9]+'), 'output-')))
+```
+
+Using this in a `WHERE` clause will result in only the files with relevant dates are scanned.
+
+#### 1.2 Counting bad rows per error:
+
+With the below SQL we can count how many bad rows per day we get for each error message:
+
+```SQL
+SELECT
+  TIMESTAMP_TRUNC(failure_tstamp, DAY),
+  e.message,
+  count(*) AS bad_row_count
+FROM bad_rows.bad_rows_external,
+UNNEST(errors) e
+WHERE DATE(PARSE_TIMESTAMP('%Y-%m-%dT%H:%M:%S', LTRIM(REGEXP_EXTRACT(_FILE_NAME, 'output-[0-9]+-[0-9]+-[0-9]+T[0-9]+:[0-9]+:[0-9]+'), 'output-'))) >= DATE_SUB(CURRENT_DATE, INTERVAL 7 DAY)
+GROUP BY 1,2
+ORDER BY 1,3 DESC
+```
+
+The `UNNEST` function is used because the errors object contains an array of `STRUCT` objects - so do note that rows with more than one error message will be counted against both messages.
+
+The output of this query looks something like this:
+
+![counts per message][counts-per-message]
+
+Some error messages will already give you an indication of what to fix - for example a schema path issue indicate that either the schema wasn't correctly uploaded to Iglu, or the tracker references the wrong path.
```

Comment:
Add "But for some, you'll also need to look at the custom data that created the error, to figure out exactly where the validation failure comes from"

Reply:
"Which we will do in the next section."